Add accessors to ComponentSelection

ljacomet opened the pull request:
The Groovy DSL gave access to ComponentMetadata and IvyModuleDescriptor
through the use of dynamic closure parameters.
These extra parameters are now available directly on the
ComponentSelection allowing access to them from the Action based API.
Note that this implies a slight semantical change, where rules using
closures would never see a null ComponentMetadata or IvyModuleDescriptor
while the API on ComponentSelection are effectively nullable.

eskatos commented on subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentSelection.java:
```diff
@@ -32,8 +35,34 @@
      *
      * @return the candidate version of the module
      */
+    @Nonnull
```
:x: `@Nonnull` should be set by default on the package level using `@NonNullApi` api instead

ljacomet replied:
And that's already done, so removing the annotation here.

eskatos commented:
With the proposed API, a kotlin-dsl script could look as follows:

```kotlin
configurations {
    create("metadataRulesConfig") {
        resolutionStrategy {
            componentSelection {
                all {
                    if(candidate.group == "org.sample" && metadata?.status == "experimental") {
                        reject("don't use experimental candidates from 'org.sample'")
                    }
                }
                withModule("org.sample:api") {
                    if (getDescriptor(IvyModuleDescriptor::class)?.branch != "release" && metadata?.status != 'milestone') {
                        selection.reject("'org.sample:api' must have testing branch or milestone status")
                    }
                }
            }
        }
    }
}
```

ljacomet commented:
That's correct. The only way to improve this, by for example having non `null` metadata and ivy module descriptor would require more work and duplicating the closure based logic where if an argument cannot be provided the rule is not executed. Not sure it is worth the effort, but happy to hear feedback from others on this.

eskatos commented:
👍 to a unified behavior

This wouldn't work as the action won't have access to the other parameters:
```
all(Action<? super ComponentSelection>, ComponentMetadata) {}
all(Action<? super ComponentSelection>, ComponentMetadata, IvyModuleDescriptor)
```

There's also the fact that because of type erasure we can't have overloads such as:
```
all(Action<? super ComponentSelection>) {}
all(Action<? super ComponentSelectionWithMetadata>) {}
all(Action<? super ComponentSelectionWithIvyModuleDescriptor>)
```

I'm probably stating the obvious here ツ

bigdaz commented:
> I'm probably stating the obvious here ツ

@eskatos Since I'm clearly missing the obvious, it's good of you to state it for me :).

ljacomet commented:
@eskatos @bigdaz I have added deprecation of the parameters detection and updated the integration tests to only use the `ComponentSelection` argument.

bigdaz commented on subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentSelection.java:
```diff
@@ -34,6 +36,31 @@
 
+    /**
+     * Gets the metadata of the component.
+     * The metadata may not be available, in which case {@code null} is returned.
```
Under what circumstances is the metadata `null`? We should document this.

ljacomet replied:
Added some information, but not 100% confident it is enough. However, looking at impl and tests, I could not find extra info on the topic. So if you know better, I am all ears 😉
